site/app/models/gradeable/GradedComponent.php

```diff
@@ -132,18 +133,33 @@ private function setComponent(Component $component) {
 
     /**
      * Sets the marks the submitter received for this component
-     * @param array $marks
+     * @param int[] $mark_ids
      */
-    public function setMarks(array $marks) {
-        $new_mark_ids = [];
-        foreach ($marks as $mark) {
-            if (!($mark instanceof Mark)) {
-                throw new \InvalidArgumentException('Object in marks array was not a mark');
+    public function setMarkIds(array $mark_ids) {
+        // This may seem redundant, but by fetching the marks from the component and calling setMarks, we
+        //  effectively filter out any of the invalid values in $mark_ids
+        $marks = [];
+        $actual_ids = [];
+        foreach ($this->component->getMarks() as $mark) {
+            if (in_array($mark->getId(), $mark_ids)) {
+                $marks[] = $mark;
+                $actual_ids[] = $mark->getId();
             }
-            $new_mark_ids[] = $mark->getId();
         }
         $this->marks = $marks;
-        $this->mark_ids = $new_mark_ids;
+        $this->mark_ids = $actual_ids;
+        $this->marks_modified = true;
+    }
+
+    /**
+     * Called from the db methods to load the mark ids the submitter received
+     * @param int[] $db_mark_ids
+     * @internal
+     */
+    public function setMarkIdsFromDb(array $db_mark_ids) {
```

Comment:
`setDbMarkIdsFromDatabase` to be consistent with `setIdFromDatabase`

Reply:
This sets both, I think adding `db` prefix would imply that it doesn't also set the `mark_ids`

Reply:
To be fair the whole db_mark_ids business seems like a messy way of dealing with the lack of a `GradedMark::$modified` type deal for knowing which haven't been saved

Reply:
Not sure if I can come up with a better solution than array_diff though